File: sdk/dotnet/Internal.Telemetry/TelemetryConfig.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Net.Http;

namespace Internal.Telemetry;

/// <summary>
/// The resolved configuration one client runs on. Built once, then read-only, so no
/// request path ever parses an environment variable.
/// </summary>
public sealed class TelemetryConfig
{
    /// <summary>The collector refuses a batch larger than this.</summary>
    public const int MaxBatch = 500;

    private TelemetryConfig(
        string service,
        string? endpoint,
        bool enabled,
        string eventsPath,
        string correlationsPath,
        int batchSize,
        int maxQueueSize,
        TimeSpan requestTimeout,
        SourceMatcher syntheticSources,
        int maxBodyBytes,
        int maxAttributes,
        int maxBodyDepth,
        bool reportDiscards,
        HttpMessageHandler? messageHandler)
    {
        Service = service;
        Endpoint = endpoint;
        Enabled = enabled;
        EventsPath = eventsPath;
        CorrelationsPath = correlationsPath;
        BatchSize = batchSize;
        MaxQueueSize = maxQueueSize;
        RequestTimeout = requestTimeout;
        SyntheticSources = syntheticSources;
        MaxBodyBytes = maxBodyBytes;
        MaxAttributes = maxAttributes;
        MaxBodyDepth = maxBodyDepth;
        ReportDiscards = reportDiscards;
        MessageHandler = messageHandler;
    }

    /// <summary>Service name stamped on every record.</summary>
    public string Service { get; }

    /// <summary>Collector base address, or null when none was configured.</summary>
    public string? Endpoint { get; }

    /// <summary>False makes every entry point a no-op.</summary>
    public bool Enabled { get; }

    /// <summary>Path batched records are posted to.</summary>
    public string EventsPath { get; }

    /// <summary>Path dependency links are posted to.</summary>
    public string CorrelationsPath { get; }

    /// <summary>Records per POST.</summary>
    public int BatchSize { get; }

    /// <summary>Records held in memory before the oldest are dropped.</summary>
    public int MaxQueueSize { get; }

    /// <summary>Budget for one export call.</summary>
    public TimeSpan RequestTimeout { get; }

    /// <summary>Compiled prefixes for the estate's own generated traffic.</summary>
    public SourceMatcher SyntheticSources { get; }

    /// <summary>Request body bytes buffered for attribute extraction.</summary>
    public int MaxBodyBytes { get; }

    /// <summary>Attributes recorded per record.</summary>
    public int MaxAttributes { get; }

    /// <summary>Nesting depth walked when flattening a JSON document.</summary>
    public int MaxBodyDepth { get; }

    /// <summary>Whether dropped records are announced as a note.</summary>
    public bool ReportDiscards { get; }

    /// <summary>Transport seam; null means an ordinary socket handler.</summary>
    public HttpMessageHandler? MessageHandler { get; }

    /// <summary>
    /// Resolve options against the environment. Explicit values always win; anything
    /// left unset falls back to a TELEMETRY_* variable and then to a default.
    /// </summary>
    public static TelemetryConfig Resolve(TelemetryOptions? options = null)
    {
        options ??= new TelemetryOptions();
        Func<string, string?> env = options.EnvironmentReader ?? Environment.GetEnvironmentVariable;

        string service = options.Service ?? env("TELEMETRY_SERVICE") ?? string.Empty;
        string endpointRaw = options.Endpoint ?? env("TELEMETRY_ENDPOINT") ?? string.Empty;
        string? endpoint = endpointRaw.Length == 0 ? null : endpointRaw.TrimEnd('/');

        // With no service name nothing is configured, so the client stays inert. That
        // keeps it quiet in local development and in unit tests with no extra wiring.
        bool enabled = options.Enabled ?? ParseBool(env("TELEMETRY_ENABLED")) ?? service.Length > 0;

        IReadOnlyList<string> cidrs = options.SyntheticCidrs ?? SplitList(env("TELEMETRY_SYNTHETIC_CIDRS"));

        int batchSize = Clamp(options.BatchSize ?? ParseInt(env("TELEMETRY_BATCH_MAX")) ?? MaxBatch, 1, MaxBatch);
        int queueMax = Math.Max(1, options.MaxQueueSize ?? ParseInt(env("TELEMETRY_QUEUE_MAX")) ?? 10_000);
        TimeSpan timeout = options.RequestTimeout
            ?? TimeSpan.FromSeconds(ParseDouble(env("TELEMETRY_TIMEOUT_S")) ?? 5.0);
        if (timeout <= TimeSpan.Zero)
        {
            timeout = TimeSpan.FromSeconds(5);
        }

        return new TelemetryConfig(
            service: service,
            endpoint: endpoint,
            enabled: enabled && service.Length > 0,
            eventsPath: WithLeadingSlash(options.EventsPath ?? env("TELEMETRY_EVENTS_PATH") ?? "/v1/traces"),
            correlationsPath: WithLeadingSlash(
                options.CorrelationsPath ?? env("TELEMETRY_CORRELATIONS_PATH") ?? "/v1/correlations"),
            batchSize: batchSize,
            maxQueueSize: queueMax,
            requestTimeout: timeout,
            syntheticSources: SourceMatcher.Compile(cidrs),
            maxBodyBytes: Math.Max(0, options.MaxBodyBytes ?? ParseInt(env("TELEMETRY_MAX_BODY_BYTES")) ?? 262_144),
            maxAttributes: Math.Max(1, options.MaxAttributes ?? ParseInt(env("TELEMETRY_MAX_PARAMS")) ?? 1024),
            maxBodyDepth: Math.Max(1, options.MaxBodyDepth ?? 16),
            reportDiscards: options.ReportDiscards ?? true,
            messageHandler: options.MessageHandler);
    }

    private static int Clamp(int value, int low, int high)
    {
        if (value < low)
        {
            return low;
        }

        return value > high ? high : value;
    }

    private static string WithLeadingSlash(string path)
    {
        if (string.IsNullOrEmpty(path))
        {
            return "/";
        }

        return path[0] == '/' ? path : "/" + path;
    }

    private static bool? ParseBool(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw))
        {
            return null;
        }

        switch (raw.Trim().ToLowerInvariant())
        {
            case "0":
            case "false":
            case "no":
            case "off":
                return false;
            case "1":
            case "true":
            case "yes":
            case "on":
                return true;
            default:
                return null;
        }
    }

    private static int? ParseInt(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw))
        {
            return null;
        }

        return int.TryParse(raw.Trim(), NumberStyles.Integer, CultureInfo.InvariantCulture, out int value)
            ? value
            : null;
    }

    private static double? ParseDouble(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw))
        {
            return null;
        }

        return double.TryParse(raw.Trim(), NumberStyles.Float, CultureInfo.InvariantCulture, out double value)
            ? value
            : null;
    }

    private static IReadOnlyList<string> SplitList(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw))
        {
            return Array.Empty<string>();
        }

        List<string> parts = new();
        foreach (string chunk in raw.Split(new[] { ',', ' ', '\t', '\n', '\r' }, StringSplitOptions.RemoveEmptyEntries))
        {
            string trimmed = chunk.Trim();
            if (trimmed.Length > 0)
            {
                parts.Add(trimmed);
            }
        }

        return parts;
    }
}

File: sdk/java/src/main/java/internal/telemetry/TelemetryConfig.java
package internal.telemetry;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * Runtime configuration, read from explicit values first and then from the environment.
 *
 * <p>Services are deployed as containers whose environment carries the {@code TELEMETRY_*}
 * keys, so the environment is the primary source; the builder exists for unit tests and
 * for services that own their own configuration loading.
 *
 * <p>Recognised variables:
 * <ul>
 *   <li>{@code TELEMETRY_SERVICE} — name reported with every record.</li>
 *   <li>{@code TELEMETRY_ENDPOINT} — collector base URL.</li>
 *   <li>{@code TELEMETRY_ENABLED} — master switch.</li>
 *   <li>{@code TELEMETRY_EVENTS_PATH} — default {@code /v1/traces}.</li>
 *   <li>{@code TELEMETRY_CORRELATIONS_PATH} — default {@code /v1/correlations}.</li>
 *   <li>{@code TELEMETRY_SYNTHETIC_CIDRS} — peer networks whose traffic is generated
 *       rather than organic (uptime probes, warm-up jobs, load generators), decided on
 *       the socket peer address and on nothing else.</li>
 *   <li>{@code TELEMETRY_QUEUE_MAX}, {@code TELEMETRY_BATCH_MAX},
 *       {@code TELEMETRY_FLUSH_INTERVAL_MS}, {@code TELEMETRY_TIMEOUT_MS},
 *       {@code TELEMETRY_MAX_BODY_BYTES}, {@code TELEMETRY_MAX_PARAMS}.</li>
 * </ul>
 */
public final class TelemetryConfig {

    /** The collector refuses a larger batch outright, so this is a ceiling, not a default. */
    public static final int BATCH_CEILING = 500;

    private final String service;
    private final String endpoint;
    private final boolean enabled;
    private final String eventsPath;
    private final String correlationsPath;
    private final int queueMax;
    private final int batchMax;
    private final Duration flushInterval;
    private final Duration timeout;
    private final SourceMatcher syntheticSources;
    private final int maxBodyBytes;
    private final int maxParams;

    private TelemetryConfig(Builder b) {
        this.service = b.service == null ? "" : b.service;
        this.endpoint = b.endpoint == null ? "" : stripTrailingSlash(b.endpoint);
        this.enabled = b.enabled != null ? b.enabled : !this.service.isEmpty();
        this.eventsPath = withLeadingSlash(b.eventsPath == null ? "/v1/traces" : b.eventsPath);
        this.correlationsPath =
                withLeadingSlash(b.correlationsPath == null ? "/v1/correlations" : b.correlationsPath);
        this.queueMax = b.queueMax > 0 ? b.queueMax : 10_000;
        this.batchMax = Math.min(b.batchMax > 0 ? b.batchMax : BATCH_CEILING, BATCH_CEILING);
        this.flushInterval = b.flushInterval == null ? Duration.ofMillis(250) : b.flushInterval;
        this.timeout = b.timeout == null ? Duration.ofSeconds(5) : b.timeout;
        this.syntheticSources = b.syntheticSources == null ? SourceMatcher.none() : b.syntheticSources;
        this.maxBodyBytes = b.maxBodyBytes > 0 ? b.maxBodyBytes : 262_144;
        this.maxParams = b.maxParams > 0 ? b.maxParams : 1024;
    }

    public static Builder builder() {
        return new Builder();
    }

    /** Read the environment. Unset and unparseable values fall back to the defaults. */
    public static TelemetryConfig fromEnvironment() {
        return fromEnvironment(System.getenv());
    }

    public static TelemetryConfig fromEnvironment(Map<String, String> env) {
        Builder b = builder()
                .service(env.get("TELEMETRY_SERVICE"))
                .endpoint(env.getOrDefault("TELEMETRY_ENDPOINT", "http://otel-collector:8900"))
                .eventsPath(env.get("TELEMETRY_EVENTS_PATH"))
                .correlationsPath(env.get("TELEMETRY_CORRELATIONS_PATH"))
                .queueMax(intOf(env.get("TELEMETRY_QUEUE_MAX"), 0))
                .batchMax(intOf(env.get("TELEMETRY_BATCH_MAX"), 0))
                .maxBodyBytes(intOf(env.get("TELEMETRY_MAX_BODY_BYTES"), 0))
                .maxParams(intOf(env.get("TELEMETRY_MAX_PARAMS"), 0))
                .syntheticCidrs(listOf(env.get("TELEMETRY_SYNTHETIC_CIDRS")));

        int flush = intOf(env.get("TELEMETRY_FLUSH_INTERVAL_MS"), 0);
        if (flush > 0) {
            b.flushInterval(Duration.ofMillis(flush));
        }
        int timeout = intOf(env.get("TELEMETRY_TIMEOUT_MS"), 0);
        if (timeout > 0) {
            b.timeout(Duration.ofMillis(timeout));
        }
        Boolean on = boolOf(env.get("TELEMETRY_ENABLED"));
        if (on != null) {
            b.enabled(on);
        }
        return b.build();
    }

    public String service() {
        return service;
    }

    public String endpoint() {
        return endpoint;
    }

    /**
     * With no service name nothing is configured, so the client stays inert and every
     * entry point is a no-op. That keeps it silent in local development and in unit
     * tests without any extra wiring.
     */
    public boolean enabled() {
        return enabled && !service.isEmpty();
    }

    public String eventsPath() {
        return eventsPath;
    }

    public String correlationsPath() {
        return correlationsPath;
    }

    public int queueMax() {
        return queueMax;
    }

    public int batchMax() {
        return batchMax;
    }

    public Duration flushInterval() {
        return flushInterval;
    }

    public Duration timeout() {
        return timeout;
    }

    public SourceMatcher syntheticSources() {
        return syntheticSources;
    }

    public int maxBodyBytes() {
        return maxBodyBytes;
    }

    public int maxParams() {
        return maxParams;
    }

    private static String stripTrailingSlash(String url) {
        String value = url.strip();
        return value.endsWith("/") ? value.substring(0, value.length() - 1) : value;
    }

    private static String withLeadingSlash(String path) {
        return path.startsWith("/") ? path : "/" + path;
    }

    private static int intOf(String raw, int fallback) {
        if (raw == null || raw.isBlank()) {
            return fallback;
        }
        try {
            return Integer.parseInt(raw.strip());
        } catch (NumberFormatException malformed) {
            return fallback;
        }
    }

    private static Boolean boolOf(String raw) {
        if (raw == null || raw.isBlank()) {
            return null;
        }
        return switch (raw.strip().toLowerCase(Locale.ROOT)) {
            case "1", "true", "yes", "on" -> Boolean.TRUE;
            case "0", "false", "no", "off" -> Boolean.FALSE;
            default -> null;
        };
    }

    private static List<String> listOf(String raw) {
        List<String> out = new ArrayList<>();
        if (raw == null) {
            return out;
        }
        for (String part : raw.split("[,\\s]+")) {
            if (!part.isBlank()) {
                out.add(part.strip());
            }
        }
        return out;
    }

    /** Mutable builder; every setter tolerates {@code null} as "leave the default". */
    public static final class Builder {
        private String service;
        private String endpoint;
        private Boolean enabled;
        private String eventsPath;
        private String correlationsPath;
        private int queueMax;
        private int batchMax;
        private Duration flushInterval;
        private Duration timeout;
        private SourceMatcher syntheticSources;
        private int maxBodyBytes;
        private int maxParams;

        public Builder service(String value) {
            this.service = value;
            return this;
        }

        public Builder endpoint(String value) {
            this.endpoint = value;
            return this;
        }

        public Builder enabled(boolean value) {
            this.enabled = value;
            return this;
        }

        public Builder eventsPath(String value) {
            this.eventsPath = value;
            return this;
        }

        public Builder correlationsPath(String value) {
            this.correlationsPath = value;
            return this;
        }

        public Builder queueMax(int value) {
            this.queueMax = value;
            return this;
        }

        public Builder batchMax(int value) {
            this.batchMax = value;
            return this;
        }

        public Builder flushInterval(Duration value) {
            this.flushInterval = value;
            return this;
        }

        public Builder timeout(Duration value) {
            this.timeout = value;
            return this;
        }

        public Builder syntheticCidrs(List<String> value) {
            this.syntheticSources = SourceMatcher.of(value == null ? List.of() : value);
            return this;
        }

        public Builder syntheticSources(SourceMatcher value) {
            this.syntheticSources = value;
            return this;
        }

        public Builder maxBodyBytes(int value) {
            this.maxBodyBytes = value;
            return this;
        }

        public Builder maxParams(int value) {
            this.maxParams = value;
            return this;
        }

        public TelemetryConfig build() {
            return new TelemetryConfig(this);
        }
    }
}
